### ISSUE TITLE
Gap between "SoFiA" and "Jhhmmss+ddmmss" in ascii catalog in WCS units

### ISSUE THREAD

**Issue by kmhess** [SoFiA-Admin/SoFiA-2#56]
When the catalog output is in WCS units (parameter.wcs = True), the "Name" entries have a space between SoFiA and the J2000 coordinate, which makes it impossible to read with astropy as an ascii table because the number of columns between the header and the data don't match.  This lead to *a lot* of discussion about the intent of the catalog outputs.

Possible solutions:
* Remove the space or replace with charater between "SoFiA" and "Jhhmmss..."
* Delete the "SoFiA" prefix
* Add an extra column header entry to match the SoFiA prefix
* Remove the ascii catalog as an output option

Other considerations (the ascii table is not meant to be the primary catalog data product):
* Include a warning at the top of the ascii file (and in user manual) that this may not have the highest precision/is not the primary catalog data product
* Always write the XML catalog file (this is the primary catalog data product)
* Any time any catalog is written, write the XML file (allows the option to turn off all catalog generation)

**Comment by SoFiA-Admin** [comment]
Thanks for raising this issue. I just figured out that there is a trivial solution to this entire problem: If we place the source name in quotation marks, then Astropy will read the catalogue correctly. For example, take the following catalogue:

```
# name   x   y   z
#
"src 1"  1.0 2.0 3.0
"src 2"  2.0 3.0 4.0
```

Saving this as `catalog.dat` and then reading that file into Astropy with

```python
from astropy.io import ascii
ascii.read("catalog.dat")
```

yields

```
<Table length=2>
 name    x       y       z   
 str5 float64 float64 float64
----- ------- ------- -------
src 1     1.0     2.0     3.0
src 2     2.0     3.0     4.0
```

which is the expected result. I tested this with an original SoFiA catalogue, and it works as expected.

My suggestion therefore is to simply put the source names in the plain-text catalogue in quotation marks which would instantly resolve this entire issue. This solution would retain the original and correct IAU source naming scheme while allowing plain-text catalogues to be directly read into Python with Astropy’s `ascii.read()` method.

**Comment by SoFiA-Admin** [comment]
I have modified the plain text catalogue by adding quotation marks around the source name, which should allow the name to be correctly imported into Python while preserving any spaces in the source name.

I also made another change by setting the unit column of dimensionless parameters to `-` (ASCII hyphen) rather than just leaving it blank. This should again help with automatic parsing of the unit strings by ensuring that the correct number of columns is extracted.

Lastly, I added a disclaimer at the beginning of the plain text catalogue to recommend reading the XML (VO Table) version instead. This should effectively address all of the issues. Please test and let me know if you encounter any additional problems.

**Comment by SoFiA-Admin** [comment]
As discussed on Monday, please test the new plain-text catalogue format and check if it is correctly parsed by the various parsing functions provided by Python libraries such as Astropy or NumPy.

**Comment by SoFiA-Admin** [comment]
I did a few more tests. Firstly, NumPy’s `loadtxt()` cannot be used with mixed-type tables, as it reads everything into a single NumPy array which must have a single, well-defined type. It is still possible, though, to read individual columns of the catalogue, e.g.

```python
import numpy as np
np.loadtxt("sofiatestcube_cat.txt", usecols=(2,3,4,5))
```

yields

```
array([[  1.      , 128.951414,  20.636432,  65.454223],
       [  2.      ,  90.619178, 102.646359,  39.337757],
       [  3.      , 113.718169, 174.973926,  22.764293]])
```

Note that everything is treated as a floating-point number, including the source ID (first column) which is originally an integer number.

The other method I tested is the Astropy `Table` module, which seems to be working well. For example

```python
from astropy.table import Table
Table.read("sofiatestcube_cat.txt", format="ascii")
```

yields

```
<Table length=3>
           col1            col2    col3    ...   col44      col45     col46  
          str25           int64  float64   ...  float64    float64   float64 
------------------------- ----- ---------- ... ---------- --------- ---------
SoFiA J120121.19+615351.4     1 128.951414 ... 180.296731 61.893037 1596750.0
SoFiA J120215.80+620729.9     2  90.619178 ... 180.522036 62.131568 1308180.0
SoFiA J120142.97+621931.6     3 113.718169 ... 180.433364 62.328277 1093820.0
```

which is the correct and expected output. The source name is correctly parsed and does include the space, as would be expected.

In summary, I think the solution of putting the source name in quotation marks is robust and seems to work well with different text file input methods in Python. Even `numpy.loadtxt()` will work, although it cannot read strings, and the first two columns with the source identifier + coordinate therefore need to be skipped. All other columns are read correctly, though, albeit as floating-point numbers.

**Comment by kmhess** [comment]
I've tested the latest version of SoFiA and I'm happy with the
implementation of the quotes.  I mentioned testing in Pandas before, but
this is going to take a bit more effort since I don't usually use it.

However--and this is maybe a small thing based on personal preference,
but--I would suggest that we change the recommended way to read in the xml
table in the User manual.  Using the "parse_single_table" seems particular
to xml, and at least for me, is not very intuitive.  This is also inspired
by trying to adapt my code to easily allow for either xml or ascii input.

If we suggest using astropy.tables, then both the xml and ascii tables can
be read into python in a consistent way, and in a way that can maintain the
column names. (Note that the header_start value is probably dependent on
the what version of SoFiA used to create the table--the latest version has
the warning about accuracy in the comments.):

(1) The xml table:

*from* *astropy.table* *import* Table

astrox = Table.read("test_data/data_cat.xml")

astrox['name']

Out[*48*]:

<MaskedColumn name='name' dtype='str32' unit='' length=6>

SoFiA B115929.29+300737.7

SoFiA B120113.71+302430.9

SoFiA B115948.69+300824.0

SoFiA B115907.06+302743.0

SoFiA B115915.01+301407.5

SoFiA B120115.37+295953.9


(2) The ascii table:

astro = Table.read("test_data/data_cat.txt",format='ascii',header_start=18)

astro['name']

Out[*49*]:

<Column name='name' dtype='str25' length=6>

SoFiA B115929.29+300737.7

SoFiA B120113.71+302430.9

SoFiA B115948.69+300824.0

SoFiA B115907.06+302743.0

SoFiA B115915.01+301407.5

SoFiA B120115.37+295953.9


As opposed to the old version:

*from* *astropy.io.votable* *import* parse_single_table

xml = parse_single_table("test_data/data_cat.xml")

xml['name']

---------------------------------------------------------------------------

TypeError                                 Traceback (most recent call last)

<ipython-input-50-a32f4e750030> in <module>

----> 1 xml['name']


TypeError: 'Table' object is not subscriptable


In [*51*]: xml.array['name']

Out[*51*]:

masked_array(data=['SoFiA B115929.29+300737.7',

                   'SoFiA B120113.71+302430.9',

                   'SoFiA B115948.69+300824.0',

                   'SoFiA B115907.06+302743.0',

                   'SoFiA B115915.01+301407.5',

                   'SoFiA B120115.37+295953.9'],

             mask=[False, False, False, False, False, False],

       fill_value='N/A',

            dtype='<U32')



As an aside, one can recreate the astropy.table format with the correct
column names using ascii.read as well, so I would postulate that
astropy.table or ascii.read will be a more common way for users to try and
access the catalog data.

*from* *astropy.io <http://astropy.io>* *import* ascii

asc = ascii.read('test_data/data_cat.txt',format='commented_header'
,header_start=18)

asc['name']

Out[*55*]:

<Column name='name' dtype='str25' length=6>

SoFiA B115929.29+300737.7

SoFiA B120113.71+302430.9

SoFiA B115948.69+300824.0

SoFiA B115907.06+302743.0

SoFiA B115915.01+301407.5

SoFiA B120115.37+295953.9

**Comment by SoFiA-Admin** [comment]
Thanks for that suggestion. I agree that we should change the recommended method for reading catalogues into Python, as the `Table` module seems more flexible and has the ability to read both the XML and ASCII formats.

Note that the original `votable` method does allow you to access columns by name:

```python
from astropy.io.votable import parse_single_table
table = parse_single_table("sofiatestcube_cat.xml").array
table["name"].data
```

yields

```
array([b'SoFiA J120121.19+615351.4', 
       b'SoFiA J120215.80+620729.9',
       b'SoFiA J120142.97+621931.6'], dtype='|S32')
```

However, this is slightly more verbose, so I’d be happy to recommend the `Table`module as the preferred method instead.

**Comment by SoFiA-Admin** [comment]
Thanks for testing Pandas. I did a few tests myself, and it seems rather difficult to get the catalogue read into Pandas. While there is a `read_table()` method that allows one to specify the header row and comment character, the way this is currently implemented in Pandas makes it unfortunately impossible to have the header line start with a comment character! There are several complaints and GitHub issues about this problem (see https://github.com/pandas-dev/pandas/issues/20378 or https://stackoverflow.com/questions/36772656/get-one-specific-line-of-comment-as-header-with-python-pandas).

In the end I found a way on simply manually providing the column widths to `read_fwf()` in the following way:

```python
pd.read_fwf("sofiatestcube_cat.txt", quotechar='"', header=18, skiprows=[19,20], colspecs=[(1,29)]+[(29+14*i,43+14*i) for i in range(45)])
```

This will parse the catalogue correctly, ignore the first character in each line (to get rid of the `#`), read the correct header row and ignore the two rows below the header row, as they don’t contain data. The resulting table should have the correct format and contain all of the valid columns without invalid rows or NaNs. The `range(45)` may need to be adjusted, though, depending on how many columns (minus one) there are in the actual catalogue file to be parsed.

Overall, I get the impression that the Pandas data import functions are rather clunky and poorly conceived, and one would be better off using the many alternative options provided by Astropy & friends.

**Comment by SoFiA-Admin** [comment]
This has been fixed and can be closed now.
